File: perl/day4_part1_2017.pl

use strict;
use warnings;

open my $fh, '<', "input.txt" or die "Cannot open file: $!";
my @passphrases = map { chomp; $_ } <$fh>;
close $fh;

my $validCount = 0;

foreach my $passphrase (@passphrases) {
    my @words = split(' ', $passphrase);
    my %wordSet;

    my $valid = 1;
    foreach my $word (@words) {
        if ($wordSet{$word}) {
            $valid = 0;
            last;
        }
        $wordSet{$word} = 1;
    }

    if ($valid) {
        $validCount++;
    }
}

print "$validCount\n";

File: dart/day4_part1_2017.dart

import 'dart:io';

void main() {
  List<String> passphrases = File('input.txt').readAsLinesSync();
  int validCount = 0;

  for (String passphrase in passphrases) {
    List<String> words = passphrase.split(' ');
    Set<String> uniqueWords = Set<String>.from(words);

    if (words.length == uniqueWords.length) {
      validCount++;
    }
  }

  print(validCount);
}
